Partial import 384

VolkerHBZ opened the pull request:
enable incomplete imports by ignoring invalid data

VolkerHBZ committed:
Change the calculation of Statistics to only count Components that are not deleted

VolkerHBZ committed:
Ignoring invalid TIPPs during crossReferencePackage

VolkerHBZ committed:
More info in the aftermath request for review

philboeselager reviewed (approved):
Code looks fine.

hornmo commented on server/gokbg3/grails-app/controllers/org/gokb/IntegrationController.groovy:
```diff
@@ -1084,17 +1091,14 @@ class IntegrationController {
-                      valid &= valid_plt?.valid
-
-                      if (!valid_plt.valid) {
+                      if (!Platform.validateDTO(tipp_plt_dto)) {
```
Platform.validateDTO returns a Map `[ valid: 'true/false', errors:[:] ]`

hornmo commented on server/gokbg3/grails-app/controllers/org/gokb/IntegrationController.groovy:
```diff
@@ -1104,11 +1108,8 @@ class IntegrationController {
-
-                      if (valid) {
-
+                        errors.tipps.add([plt])
```
Should be `plt_errors` instead of `plt`, failed platform validation invalidates the TIPP.

hornmo commented on server/gokbg3/grails-app/controllers/org/gokb/IntegrationController.groovy:
```diff
@@ -1156,7 +1157,7 @@ class IntegrationController {
                         log.warn("No package");
                         errors.tipps.add(['code': 400, idx: idx, 'message': messageService.resolveCode('crossRef.package.tipps.error.pkgId', [json_tipp.title.name], locale)])
-                        valid = false
+                        allTippsValid = false
```
Has to be  added to invalidTIPPs.

hornmo commented on server/gokbg3/grails-app/controllers/org/gokb/IntegrationController.groovy:
```diff
@@ -1125,13 +1126,13 @@ class IntegrationController {
                             log.error("ValidationException attempting to cross reference title", ve);
                             valid_plt = false
-                            valid = false
+                            allTippsValid = false
```
Has to be  added to invalidTIPPs.

hornmo commented on server/gokbg3/grails-app/controllers/org/gokb/IntegrationController.groovy:
```diff
@@ -1125,13 +1126,13 @@ class IntegrationController {
                               log.error("Could not find/create ${tipp_plt_dto}")
                               errors.tipps.add(['code': 400, idx: idx, 'message': messageService.resolveCode('crossRef.package.tipps.error.platform', [tipp_plt_dto.name], locale)])
-                              valid = false
+                              allTippsValid = false
```
Has to be  added to invalidTIPPs.

hornmo commented on server/gokbg3/grails-app/controllers/org/gokb/IntegrationController.groovy:
```diff
@@ -1175,34 +1176,36 @@ class IntegrationController {
                       idx++
                       job.setProgress(idx, json.tipps.size() * 2)
                     }
+                    json.tipps.removeAll(invalidTipps)
```
Remove by index? Adding an 'invalid' flag to the JSON object and skipping based on that in later stages might be be an alternative?

VolkerHBZ replied:
If we enhance the data model here, it probably makes sense to keep track of the different possibilities to fail for the process and the necessity to pipe that information into a specific answer. 
This needs further discussion before implementation I'd say.
For the time being, this works.

hornmo commented on server/gokbg3/grails-app/controllers/org/gokb/IntegrationController.groovy:
```diff
@@ -1250,8 +1247,8 @@ class IntegrationController {
                     }
-
-                    // If valid, upsert tipps
+                    json.tipps.removeAll(invalidTipps)
```
See above for handling invalid TIPPs

hornmo commented on server/gokbg3/grails-app/controllers/org/gokb/IntegrationController.groovy:
```diff
@@ -1371,93 +1367,89 @@ class IntegrationController {
                     }
+
+                    additionalInfo.vars = [job.id]
+                    additionalInfo.otherComponents = invalidTipps
```
Since it is used as a variable in the Request message, `ìnvalidTipps.size()`has to be added to `additionalInfo.vars` as well.
Since these JSON-Objects have been explicitly not referenced/persisted, their usage as 'components' is problematic, since this would imply they would be able to be looked up by ID for example.

hornmo commented on server/gokbg3/grails-app/controllers/org/gokb/IntegrationController.groovy:
```diff
@@ -1524,27 +1517,27 @@ class IntegrationController {
         result.result = 'ERROR'
         response.setStatus(400)
-        result.errors = pkg_validation.errors
+        result.errors.global << pkg_validation.errors
```
The distinction between `global` and `tipp` errors has only been defined in the context of the Job processing, while `result.errors` are on a request level.

VolkerHBZ replied:
I am seriously confused by the different result and thread levels in this method. thanks for clearing that up to me.

VolkerHBZ committed:
requested changes
- Platform.validateDTO(tipp_plt_dto).valid
- result.errors vs. job_result.errors
- invalidate current tipp in exceptions

hornmo commented on server/gokbg3/grails-app/controllers/org/gokb/IntegrationController.groovy:
```diff
@@ -1371,93 +1369,89 @@ class IntegrationController {
-                      if (errors.global.size() > 0 || errors.tipps.size() > 0) {
-                        additionalInfo.errorObjects = errors
-                      }
+                  if (the_pkg) {
```
This needs to check for `invalidTipps.size() > 0` as well, otherwise there will be a RR concerning invalid TIPPs for every package import.

VolkerHBZ replied:
Sorry for reformatting the whole file. This way, the `if ... else `clauses are easier to fold away in IntelliJ.

VolkerHBZ committed:
requested change
checking invalidTipps.size before adding it to the ReviewRequest.

hornmo committed:
fix check & adjust template
